solarforecastarbiter/pvmodel.py

```diff
@@ -0,0 +1,428 @@
+        surface_tilt, surface_azimuth,
+        dhi, dni, dni_extra,
+        solar_zenith=apparent_zenith,
+        solar_azimuth=azimuth,
+        dni_extra=dni_extra, model='haydavies')
+    # assumes albedo = 0.25
+    poa_ground_diffuse = pvlib.irradiance.get_ground_diffuse(surface_tilt, ghi)
+    aoi_modifier = pvlib.pvsystem.physicaliam(aoi)
+    beam_component = dni * pvlib.tools.cosd(aoi)
+    beam_component = beam_component.clip(lower=0) * aoi_modifier
+    poa_effective = beam_component + poa_sky_diffuse + poa_ground_diffuse
+    return poa_effective
+
+
+def calculate_poa_effective_func(aoi_func, apparent_zenith, azimuth,
+                                 ghi, dni, dhi):
+    """
+    Calculate effective plane of array irradiance from system metadata,
+    solar position, and irradiance components. Accounts for AOI losses.
+
+    Parameters
+    ----------
+    aoi_func : function
+        Function with arguments (apparent_zenith, azimuth) and returns
+        surface_tilt, surface_azimuth, aoi
+    apparent_zenith : pd.Series
+        Solar apparent zenith
+    azimuth : pd.Series
+        Solar azimuth
+    ghi : pd.Series
+    dni : pd.Series
+    dhi : pd.Series
+
+    Returns
+    -------
+    poa_effective : pd.Series
+    """
+    surface_tilt, surface_azimuth, aoi = aoi_func(apparent_zenith, azimuth)
+    dni_extra = pvlib.irradiance.get_extra_radiation(apparent_zenith.index)
+    poa_sky_diffuse = pvlib.irradiance.haydavies(
+        surface_tilt, surface_azimuth,
+        dhi, dni, dni_extra,
+        solar_zenith=apparent_zenith,
+        solar_azimuth=azimuth,
+        dni_extra=dni_extra, model='haydavies')
+    # assumes albedo = 0.25
+    poa_ground_diffuse = pvlib.irradiance.get_ground_diffuse(surface_tilt, ghi)
+    aoi_modifier = pvlib.pvsystem.physicaliam(aoi)
+    beam_component = dni * pvlib.tools.cosd(aoi)
+    beam_component = beam_component.clip(lower=0) * aoi_modifier
+    poa_effective = beam_component + poa_sky_diffuse + poa_ground_diffuse
+    return poa_effective
+
+
+def calculate_power(dc_capacity, dc_loss_factor, ac_capacity, ac_loss_factor,
+                    poa_effective, temp_air=20, wind_speed=1):
+    """
+    Calcuate AC power from system metadata, plane of array irradiance,
+    and weather data using the PVWatts model.
+
+    Parameters
+    ----------
+    dc_capacity : float
+    dc_loss_factor : float
+    ac_capacity : float
+    ac_loss_factor : float
+    poa_effective : pd.Series
+    temp_air : pd.Series, default 20
+    wind_speed : pd.Series, default 1
+
+    Returns
+    -------
+    ac_power : pd.Series
+    """
+    pvtemps = pvlib.pvsystem.sapm_celltemp(poa_effective,
+                                           wind_speed, temp_air)
+    dc = pvlib.pvsystem.pvwatts_dc(poa_effective, pvtemps['temp_cell'],
+                                   dc_capacity)
+    dc *= (1 - dc_loss_factor / 100)
+    ac = pvlib.pvsystem.pvwatts_ac(dc, dc_capacity)
+    ac = ac.clip(upper=ac_capacity)
```

Comment:
Hmm.  we're clipping the AC output twice, once inside the PVWatts model and again here.  We may want to vary from the PVWatts default parameters to turn off clipping in PVWatts, and apply the clipping here, so that we don't later grapple with `ac_capacity = 1.0` but `ac_power` never reaches 1.0.  If you agree, I'd set both `eta_inv_nom=1.0` and `eta_inv_ref=1.0`.